File: acmicpc/acmicpc02580.py
import sys

# sys.stdin = open('acmicpc/sample.txt', 'r')

square = []
P = []
for i in range(9):
    temp = list(map(int, sys.stdin.readline().split()))
    for j in range(9):
        if temp[j] == 0:
            P.append([i, j])
    square.append(temp)

def verify(k, idx):
    H = True
    for i in range(9):
        if square[P[k][0]][i] == idx:
            H = False
            return H
    
    V = True
    for i in range(9):
        if square[i][P[k][1]] == idx:
            V = False
            return V

    S = True
    r = 3 * (P[k][0] // 3)
    c = 3 * (P[k][1] // 3)

    for i in range(3):
        for j in range(3):
            if square[r+i][c+j] == idx:
                S = False
                return S
    
    result = H & V & S
    
    return result

def solve(k):
    # base case
    if k >= len(P):
        for i in range(9):
            print(*square[i], sep=' ')
        exit(0)

    for i in range(1, 10):
        if verify(k, i): # 어떤 조건을 만족할 때, 
            square[P[k][0]][P[k][1]] = i
            solve(k+1)
            square[P[k][0]][P[k][1]] = 0 # back tracking을 위해서 원래 값을 되돌려 놓음 


solve(0)

File: acmicpc/acmicpc02580.cpp
#include <iostream>

using namespace std;

int square[9][9];
int P[81][2];
int np;
bool isSolved;

bool verify(int k, int idx) {
    bool H = true;
    for (int i=0; i < 9; i++) {
        if (square[P[k][0]][i] == idx) {
            H = false;
            return H;
        }
    }

    bool V = true;
    for (int i=0; i < 9; i++) {
        if (square[i][P[k][1]] == idx) {
            V = false;
            return V;
        }
    }

    bool S = true;
    int r = 3 * (P[k][0] / 3);
    int c = 3 * (P[k][1] / 3);
    for (int i=0; i < 3; i++) {
        for (int j=0; j < 3; j++) {
            if (square[r+i][c+j] == idx) {
                S = false;
                return S;
            }
        }
    }

    bool result = H && V && S;
    return result;
}

void solve(int k) {
    if (k >= np) {
        for (int i=0; i < 9; i++) {
            for (int j=0; j < 9; j++) {
                cout << square[i][j] << " ";
            }
            cout << "\n";
        }
        isSolved = true;
    }

    for (int i=1; i < 10; i++) {
        if (isSolved) {
            return;
        }
        if (verify(k, i)) {
            square[P[k][0]][P[k][1]] = i;
            solve(k+1);
            square[P[k][0]][P[k][1]] = 0;
        }
    }
}

int main() {
    // freopen("sample.txt", "r", stdin);
    np = 0;
    isSolved = false;

    for (int i=0; i < 9; i++) {
        for (int j=0; j < 9; j++) {
            cin >> square[i][j];
            if (square[i][j] == 0) {
                P[np][0] = i;
                P[np][1] = j;
                np++;
            }
        }
    }

    solve(0);

    return 0;
}